near-sdk-macros/src/lib.rs

```diff
@@ -145,26 +164,58 @@ pub fn near_bindgen(attr: TokenStream, item: TokenStream) -> TokenStream {
             #metadata
             #metadata_impl_gen
         })
-    } else if let Ok(mut input) = syn::parse::<ItemImpl>(item) {
+    } else if let Some(impl_block) = process_impl_block(item, false) {
+        match impl_block {
+            Ok(res) => res,
+            Err(res) => res,
+        }
+        .into()
+    } else {
+        TokenStream::from(
+            syn::Error::new(
+                Span::call_site(),
+                "near_bindgen can only be used on struct or enum definition and impl sections.",
+            )
+            .to_compile_error(),
+        )
+    }
+}
+
+// This function deals with impl block processing, generating wrappers and ABI.
+//
+// # Arguments
+// * block - impl block to process.
+// * internal - whether or not this is an internal call, if the call is internal - reserved contract
+// method checks will be skipped.
+//
+// Returns None in case the given block is not an impl block. The underlying Result has a
+// TokenStream error type, because those need to be propagated to the compiler.
+fn process_impl_block(
+    block: TokenStream,
```

Comment:
Can we pass `ItemImpl` type here to avoid one level of nesting and also a need to return `Option<>`?